tools/mirror/src/key_mapping.rs

```diff
@@ -80,28 +83,81 @@ fn map_secp256k1(
 }
 
 // This maps the public key to a secret key so that we can sign
-// transactions on the target chain.  If secret is None, then we just
+// transactions on the target chain. If secret is None, then we just
 // use the bytes of the public key directly, otherwise we feed the
 // public key to a key derivation function.
-pub fn map_key(key: &PublicKey, secret: Option<&[u8; crate::secret::SECRET_LEN]>) -> SecretKey {
-    match key {
-        PublicKey::ED25519(k) => SecretKey::ED25519(map_ed25519(k, secret)),
-        PublicKey::SECP256K1(k) => SecretKey::SECP256K1(map_secp256k1(k, secret)),
+// If the `key` corresponds to a ETH-implicit `account_id`,
+// then the new key will be derived from that ETH address.
+pub fn map_key(
+    key: &PublicKey,
+    secret: Option<&[u8; crate::secret::SECRET_LEN]>,
+    account_id: &AccountId,
+) -> SecretKey {
+    match account_id.get_account_type() {
+        AccountType::NearImplicitAccount => {
+            match key {
+                PublicKey::ED25519(k) => SecretKey::ED25519(map_ed25519(k, secret)),
+                // TODO Is it true?
+                _ => panic!("NEAR-implicit account can only have ED25519 access key added"),
+            }
+        },
+        AccountType::EthImplicitAccount => {
+            match key {
+                PublicKey::SECP256K1(_) => map_secp256k1_from_eth_address(account_id, secret),
+                // TODO Is it true?
+                _ => panic!("ETH-implicit account can only have SECP256K1 access key added"),
+            }
+        },
+        AccountType::NamedAccount => {
+            match key {
+                PublicKey::ED25519(k) => SecretKey::ED25519(map_ed25519(k, secret)),
+                PublicKey::SECP256K1(k) => SecretKey::SECP256K1(map_secp256k1(k, secret)),
+            }
+        }
     }
 }
 
-// If it's an implicit account, interprets it as an ed25519 public key, maps that and then returns
-// the resulting implicit account. Otherwise does nothing. We do this so that transactions creating
-// an implicit account by sending money will generate an account that we can control
+// TODO Deriving secret keys from ETH addresses reduces potential entropy.
+// However, it is only for the Mirror tool. Are we ok with that?
```

Comment:
ETH address is 20 bytes length, shorter than public key (32 bytes) from which it is derived.
It reduces potential entropy so it has security impact.
But `mirror` needs to map ETH address to a new one, and we first need to map the address to a new public key, from which we then derive a new ETH address.
Are we ok with the security impact for the `mirror` tool?